Python/traceback.c

```diff
@@ -517,9 +532,54 @@ tb_displayline(PyObject *f, PyObject *filename, int lineno, PyObject *name)
     Py_DECREF(line);
     if (err != 0)
         return err;
+    int truncation = _TRACEBACK_SOURCE_LINE_INDENT;
+    PyObject* source_line = NULL;
     /* ignore errors since we can't report them, can we? */
-    if (_Py_DisplaySourceLine(f, filename, lineno, 4))
+    if (!_Py_DisplaySourceLine(f, filename, lineno, _TRACEBACK_SOURCE_LINE_INDENT, &truncation, &source_line)) {
+        int code_offset = tb->tb_lasti;
+        if (PyCode_Addr2Line(_PyFrame_GetCode(frame), code_offset) != _PyCode_Addr2EndLine(_PyFrame_GetCode(frame), code_offset)) {
+            goto done;
+        }
+
+        Py_ssize_t start_offset = (Py_ssize_t) _PyCode_Addr2Offset(_PyFrame_GetCode(frame), code_offset);
+        Py_ssize_t end_offset = (Py_ssize_t) _PyCode_Addr2EndOffset(_PyFrame_GetCode(frame), code_offset);
```

Comment:
Even though they are relatively cheap, why are we keep making `_PyFrame_GetCode` calls? (here as well as in the PyCode_Addr2Line). Also can't we just use `PyCode_Addr2Location` and get everything in one go (at worst case, it will cost 2 additional calls, but I think it worths to less complexity).

Reply:
Changed to one `_PyFrame_GetCode` and `PyCode_Addr2Location` call.